test/functional/connection.test.js

```diff
@@ -514,4 +515,27 @@ describe('Connection', function() {
       });
     }
   });
+
+  it('should be able to connect again after close', function() {
+    return withClient(this.configuration.newClient(), client => done => {
```

Comment:
I don't think we want a version of `withClient` that does it this way, returning functions which return functions can be very confusing for readability. I think it would be more useful for developers if we had something like:

```
withClient(this.configuration.newClient(), (client, done) => {

});
```

`withClient` itself could still return a `Promise`, you'd just have to determine the arity of the lambda passed in to see if a `done` callback was provided.